### ISSUE TITLE
pysam installation is unstable 

### ISSUE THREAD

**Issue by dereneaton** [dereneaton/ipyrad#252]
Finding a working pysam installation is crazy confusing. The main problem seems to be that conda *sometimes* requires a system gcc compiler, as opposed to its own, and this causes some library installations to fail on systems with older compilers like centos/redhat. 

The problem that I keep seeing is `ImportError: libhts.so.1: cannot open shared object file: No such file or directory`, and you can find a bunch of people have found this error for other reasons online as well. Generally they all say to just install some new libraries but there doesn't seem to be a really easy fix for conda. 

The pysam conda recipe seems to take a hackish approach by adding a `sed` command into its `build.sh` scripts to modify the code to avoid this error. BUT, it doesn't seem to work. At least not on the newer versions. I hadn't noticed this problem for quite a while b/c you don't see the error if you are on a newer version of ubuntu, which is what most of my systems are running. 

One way to get around this error, but which is super hackish is to install an older version of pysam/htslib/samtools/bcftools from bioconda. Very old versions (like v.0.6) lack some of the functions we need, so don't go that far back. But v.0.10.0 seems to work. So for now a working solution is:
```
conda install -c bioconda pysam=0.10.0 -f
conda install -c ipyrad ipyrad=0.7.2
```

BUT, It's not really kind of us to enforce that users need to have an old version of samtools installed in there conda environment to use ipyrad. So I would like to get it working with a very new version of pysam, but one that we keep fixed on our channel.

**Comment by isaacovercast** [comment]
To reproduce:
* Clean install of miniconda
* conda install -c ipyrad ipyrad
* $python
* import pysam
```
>>> import pysam
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/scratch/isaac.overcast/miniconda2/lib/python2.7/site-packages/pysam/__init__.py", line 5, in <module>
    from pysam.libchtslib import *
ImportError: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by /scratch/isaac.overcast/miniconda2/lib/python2.7/site-packages/pysam/libchtslib.so)
>>> 
```

**Comment by isaacovercast** [comment]
For some reason this manifests in different ways:
```
In [1]: import pysam
ImportError Traceback (most recent call last)
<ipython-input-1-9b87d3f66c91> in <module>()
----> 1 import pysam
/home/jclugston/miniconda2/lib/python2.7/site-packages/pysam/init.py in <module>()
3 import sysconfig
4
----> 5 from pysam.libchtslib import 
6 from pysam.libcutils import 
7 import pysam.libcutils as libcutils
ImportError: /home/jclugston/miniconda2/lib/././libcom_err.so.3: symbol k5_strerror_r, version krb5support_0_MIT not defined in file libkrb5support.so.0 with link time reference
```

**Comment by isaacovercast** [comment]
Current workaround is to completely remove stuff an install pysam by hand.
```
conda remove ipyrad pysam samtools bcftools htslib 
conda install pysam=0.10.0 -c bioconda
conda install ipyrad -c ipyrad -f
```

**Comment by isaacovercast** [comment]
Now getting the kerberos error seen above. This is a helpful link: https://groups.google.com/a/continuum.io/forum/#!topic/anaconda/2-bXTbSiQzg

If LD_LIBRARY_PATH is set then conda will try to use the local version of the krb5 library which is fsck. Can fix by unsetting LD_LIBRARY_PATH:
```
isaac.overcast@andy:~> export LD_LIBRARY_PATH=""
isaac.overcast@andy:~> echo $LD_LIBRARY_PATH

isaac.overcast@andy:~> python
Python 2.7.13 |Continuum Analytics, Inc.| (default, Dec 20 2016, 23:09:15) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> import pysam
>>>
```

**Comment by isaacovercast** [comment]
According to pysam docs you can choose to either use bundled htslib or external. 

https://github.com/pysam-developers/pysam/blob/v0.9.0/setup.py#L79

Commenting out a few lines in pysam/build.sh will force pysam to build samtools/bcftools/htslib, so we don't have to keep these in the repo, we will only have to build and provide pysam in the ipyrad channel:
```
#export HTSLIB_LIBRARY_DIR=$PREFIX/lib
#export HTSLIB_INCLUDE_DIR=$PREFIX/include
$PYTHON setup.py install
```

The trick is just to build it on the oldest fuckin box you can get your hands on (for linux).

**Comment by isaacovercast** [comment]
Mac build still broken 0.7.2

**Comment by isaacovercast** [comment]
I built the linux version of pysam on this box:
```
acuser@ada:ipyrad$ ldd --version
ldd (Debian EGLIBC 2.11.3-4+deb6u11) 2.11.3
```

**Comment by isaacovercast** [comment]
Fixed on mac. Fcuk my life, this was one of the nastiest sets of problems ever. After the linux build got working the mac build still was broken because the patch file was malformed so it wasn't applying (actually the file that the patch file was applying to was different than the one it was generated for). Anyway! Took for foucking ever to debug because the behavior did not scream "Failed the build!" but more like "magics is not hapnenning".... 

Long story short, the pysam people were doing some weird shit with the `@rpath` variable that describes where libraries should be loaded with that just wasn't operating right outside the context of their own install. I found that if you replace their `@rpath` magics with `@loader_path` which is the default conda thinger, then it "just works".

**Comment by isaacovercast** [comment]
This ended up being helpful:

https://conda.io/docs/building/shared-libraries.html
